src/main/java/org/folio/rest/util/CustomFieldsManager.java

```diff
@@ -0,0 +1,151 @@
+import static org.folio.rest.util.UserImportAPIConstants.IDLE_TO;
+import static org.folio.rest.util.UserImportAPIConstants.OKAPI_MODULE_ID_HEADER;
+import static org.folio.rest.util.UserImportAPIConstants.PUT_CUSTOM_FIELDS_ENDPOINT;
+import static org.folio.rest.util.UserImportAPIConstants.USERS_INTERFACE_NAME;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import io.vertx.core.Future;
+import io.vertx.core.Vertx;
+import io.vertx.core.http.HttpMethod;
+import io.vertx.core.json.JsonObject;
+
+import org.apache.commons.collections4.map.CaseInsensitiveMap;
+
+import org.folio.rest.jaxrs.model.CustomFields;
+import org.folio.rest.jaxrs.model.User;
+import org.folio.rest.jaxrs.model.UserdataimportCollection;
+import org.folio.rest.tools.client.HttpClientFactory;
+import org.folio.rest.tools.client.Response;
+import org.folio.rest.tools.client.interfaces.HttpClientInterface;
+import org.folio.rest.tools.utils.TenantTool;
+
+public final class CustomFieldsManager {
+
+  private CustomFieldsManager() {
+    throw new UnsupportedOperationException("Util class");
+  }
+
+  public static Future<Void> checkAndUpdateCustomFields(Map<String, String> okapiHeaders,
+                                                        UserdataimportCollection userCollection, Vertx vertx) {
+    Future<Void> future = Future.future();
+    Map<String, Set<String>> customFieldsOptions = getCustomFieldsOptions(userCollection);
+
+    Map<String, String> headers = new CaseInsensitiveMap<>(okapiHeaders);
+    HttpClientInterface httpClient = HttpClientFactory
+      .getHttpClient(getOkapiUrl(okapiHeaders), -1, TenantTool.tenantId(headers),
+        true, CONN_TO, IDLE_TO, false, 30L);
+    if (customFieldsOptions.isEmpty()) {
+      future.complete();
+    } else {
+      OkapiUtil.getModulesProvidingInterface(USERS_INTERFACE_NAME, headers, vertx)
+        .compose(moduleIds -> updateHeaders(moduleIds, headers))
+        .compose(o -> requestCustomFieldsDefinitions(httpClient, headers))
+        .compose(jsonObjects -> updateCF(jsonObjects, customFieldsOptions, httpClient, headers))
+        .setHandler(o -> {
+          httpClient.closeClient();
+          if (o.succeeded()) {
+            future.complete();
+          } else {
+            future.fail(o.cause());
+          }
+        });
+    }
+    return future;
+  }
+
+  @SuppressWarnings("unchecked")
+  private static Map<String, Set<String>> getCustomFieldsOptions(UserdataimportCollection userCollection) {
+    Map<String, Set<String>> customFieldsOptions = new HashMap<>();
+    List<CustomFields> customFields = userCollection.getUsers()
+      .stream()
+      .map(User::getCustomFields)
+      .filter(Objects::nonNull)
+      .collect(Collectors.toList());
+    for (CustomFields customField : customFields) {
```

Comment:
please optimize the code to iterate ones instead of collection into the list and iterating over it